File: dungeons-26.1.2/src/main/java/fishmod/features/dungeon/SimonSaysTracker.java
package fishmod.features.dungeon;

import fishmod.features.FishHudEditor;
import fishmod.utils.Misc;
import fishmod.utils.config.values.FishSettings;
import fishmod.utils.dungeon.Phase;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.fabricmc.fabric.api.client.rendering.v1.hud.HudElementRegistry;
import net.minecraft.client.DeltaTracker;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.core.BlockPos;
import net.minecraft.resources.Identifier;
import net.minecraft.network.chat.Component;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Blocks;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Tracks Goldor (F7 P3) Simon Says rounds via block scanning.
 *
 * Detection uses a FIXED world-space box around the device: the player must be inside
 * {@code DEVICE_BOX} to lock the scan onto {@code DEVICE_CENTER}, and from there we count
 * demo "flashes" (lit sea-lantern rising edges).
 *
 * Announcing: on the FIRST light of each new demo, the rounds-completed count = the longest
 * sequence shown so far. A failed round re-shows the same/shorter sequence, so the max length
 * doesn't grow and (with the dedupe) nothing extra is sent. 5/5 comes from the in-game
 * "<you> completed a device!" message / completion title.
 *
 * Breaking: the fixed obsidian/button columns behind the device (same signal NoammAddons' SS
 * solver uses) are watched for a reset — all buttons going to air after the device was active.
 * On a break, round tracking resets to 0 and scanning goes fully quiet until the device is
 * active again, so the next demo re-announces cleanly from 1/5.
 */
public final class SimonSaysTracker {

    private SimonSaysTracker() {}

    private static final int SCAN_RADIUS  = 7;   // lit-cell box around the locked device center
    private static final long BURST_GAP_MS = 550L;

    // Fixed detection box around the Goldor SS device, from measured corner coords, extended
    // upward a few blocks so the whole player (standing or jumping) counts as "at the device".
    private static final net.minecraft.world.phys.AABB DEVICE_BOX = new net.minecraft.world.phys.AABB(
            106.65, 120.0, 92.70,
            110.70, 126.0, 95.30
    );
    private static final BlockPos DEVICE_CENTER = new BlockPos(108, 120, 94);

    // Break/restart detection (same approach as NoammAddons' SS solver): a fixed obsidian
    // column behind the device and the button column in front of it. Any obsidian cell not
    // being obsidian = the device is "active" (mid demo/attempt). Once that settles for
    // BREAK_COOLDOWN_TICKS and every button cell reads air, the device has reset — a break.
    private static final int DEV_BUTTONS_X = 110;
    private static final int DEV_OBSIDIAN_X = 111;
    private static final int DEV_Y_MIN = 120, DEV_Y_MAX = 123;
    private static final int DEV_Z_MIN = 92,  DEV_Z_MAX = 95;
    private static final int BREAK_COOLDOWN_TICKS = 12;

    private static int  round         = 0;   // completed-round count shown on the HUD (0..5)
    private static int  maxLen        = 0;   // longest demo sequence length seen this run
    private static int  lastAnnounced = 0;   // highest count already sent (dedupe)
    private static int  burstFlashes  = 0;   // rising-edge flashes in the current demo
    private static long lastFlashMs   = 0L;
    private static long lastAtDeviceMs = 0L; // last tick the player was at the device
    private static boolean primed     = false;
    private static boolean completed  = false; // SS done this run — ignore everything until next run
    private static boolean armed      = false; // Goldor's intro line seen — scanning starts here
    private static int  breakTicks    = 0;     // cooldown before an "inactive" reading can count as a break
    private static boolean canBreak   = false; // device has been seen active since the last break
    private static boolean broken     = false; // device just reset — fully off (no scan/announce) until it restarts
    private static boolean inP3       = false; // HUD only
    private static boolean atDevice   = false;
    private static net.minecraft.core.BlockPos deviceCenter = null;
    private static long doneAtMs = 0L; // when 5/5 fired — HUD unrenders 2s later
    private static final java.util.HashSet<Long> litPrev = new java.util.HashSet<>();

    // Reads "Simon Says: N/5" out of party chat so the HUD also registers when SOMEONE ELSE
    // does SS (we can't block-scan their device — but their mod announces to party chat).
    private static final Pattern SS_CHAT = Pattern.compile("Simon Says: (\\d)/5");

    // Goldor's intro line — arms scanning so we don't watch the device box before P3 starts.
    private static final String GOLDOR_INTRO = "who dares trespass into my domain";

    // ── debug: log every block transition in a cube around the player (/ssdbg) ──
    public static boolean debug = false;
    private static final int DBG_R = 6;
    private static final java.util.HashMap<Long, net.minecraft.world.level.block.Block> dbgPrev = new java.util.HashMap<>();

    public static void init() {
        // New run (entering/leaving the dungeon) → reset everything.
        fishmod.utils.events.Events.ON_LOCATION_CHANGE.register(loc -> { reset(); return false; });

        // "<you> completed a device! (x/7) (time | time)" → our SS finish (5/5). Use the mod's own
        // game-message event (the same hook party commands use) for reliability.
        fishmod.utils.events.Events.ON_GAME_MESSAGE.register(message -> {
            if (!FishSettings.simonSaysEnabled) return false;
            String s = message.getString().replaceAll("§.", "");

            // Goldor's spawn line — start scanning the device box from here on.
            if (!armed && s.toLowerCase().contains(GOLDOR_INTRO)) {
                armed = true;
                if (debug) log("armed (Goldor intro seen)");
            }

            // Pick up "Simon Says: N/5" from party chat (ours echoed back, or a teammate's) so the
            // HUD shows progress even when WE aren't the one at the device.
            Matcher ss = SS_CHAT.matcher(s);
            if (ss.find()) {
                int n = ss.group(1).charAt(0) - '0';
                if (n >= 1 && n <= 5) {
                    round = n;
                    if (n >= 5) doneAtMs = System.currentTimeMillis();
                }
            }

            if (debug && s.contains("device")) log("msg: \"" + s + "\"");
            if (!s.contains("completed a device")) return false;
            // Must be OUR completion (teammates' device completions also broadcast). Match the name
            // loosely (anywhere in the line) so color-code spacing can't break it.
            Minecraft mc = Minecraft.getInstance();
            String self = (mc.player != null) ? mc.player.getGameProfile().name() : null;
            boolean mine = (self == null) || s.contains(self);
            if (debug) log("completed-a-device match; self=" + self + " mine=" + mine + " completed=" + completed);
            if (mine) tryComplete();
            return false;
        });

        ClientTickEvents.END_CLIENT_TICK.register(SimonSaysTracker::debugTick);
        ClientTickEvents.END_CLIENT_TICK.register(SimonSaysTracker::tick);
        HudElementRegistry.addLast(Identifier.fromNamespaceAndPath("fishmod", "simon_says_tracker"), SimonSaysTracker::renderHud);

        FishHudEditor.register("Simon Says",
                () -> FishSettings.simonSaysHudX, v -> FishSettings.simonSaysHudX = v,
                () -> FishSettings.simonSaysHudY, v -> FishSettings.simonSaysHudY = v,
                110, 14,
                () -> FishSettings.simonSaysHudScale, v -> FishSettings.simonSaysHudScale = v,
                () -> FishSettings.simonSaysHudEnabled && round > 0);
    }

    private static void tick(Minecraft client) {
        if (!FishSettings.simonSaysEnabled || client.player == null || client.level == null) {
            inP3 = false; atDevice = false; deviceCenter = null; reset(); return;
        }
        inP3 = safeInP3();

        // Once SS is done this run, ignore everything until the next run (location change).
        if (completed) { atDevice = false; return; }

        // Don't watch the device box until Goldor's intro line has been seen this run.
        if (!armed) { atDevice = false; return; }

        tickBreakState(client.level);
        // Device just broke — full shutoff. No scanning, no announcing, until it restarts.
        if (broken) {
            atDevice = false; deviceCenter = null; primed = false; burstFlashes = 0; litPrev.clear();
            return;
        }

        long now = System.currentTimeMillis();

        // Anyone (not just us — a teammate may be the one doing SS) standing at the fixed device box?
        atDevice = false;
        for (net.minecraft.world.entity.player.Player p : client.level.players()) {
            if (p.getBoundingBox().intersects(DEVICE_BOX)) { atDevice = true; break; }
        }
        if (atDevice) lastAtDeviceMs = now;

        if (!atDevice) {
            deviceCenter = null; primed = false; burstFlashes = 0; litPrev.clear();
            return;
        }

        if (deviceCenter == null) {
            deviceCenter = DEVICE_CENTER;
            primed = false; burstFlashes = 0; litPrev.clear();
            if (debug) log("locked device center " + deviceCenter.toShortString());
        }

        java.util.HashSet<Long> cur = new java.util.HashSet<>();
        scanLitCells(client.level, deviceCenter, cur);

        // Prime on the first scan so the always-lit decorative frame lanterns aren't miscounted.
        if (!primed) { litPrev.clear(); litPrev.addAll(cur); primed = true; return; }

        int newlyLit = 0;
        for (long p : cur) if (!litPrev.contains(p)) newlyLit++;
        litPrev.clear();
        litPrev.addAll(cur);

        if (newlyLit > 0) {
            // A gap before this light = a NEW demo just started. Announce here, on the FIRST light.
            if (now - lastFlashMs > BURST_GAP_MS) {
                if (burstFlashes > maxLen) maxLen = burstFlashes; // finalize previous demo
                burstFlashes = 0;
                int done = Math.min(5, maxLen);
                if (done >= 1 && done <= 4 && done > lastAnnounced) {
                    lastAnnounced = done;
                    round = done;
                    announceRound(done);
                }
            }
            burstFlashes += newlyLit;
            lastFlashMs = now;
            if (debug) log("flash +" + newlyLit + " burst=" + burstFlashes + " maxLen=" + maxLen);
        }
    }

    /**
     * 5/5 finish, triggered by the in-game "<you> completed a device!" message (the reliable
     * signal — block detection of rounds can miss). Fires once per run; the run reset clears it.
     */
    private static void tryComplete() {
        if (debug) log("tryComplete called (completed=" + completed + ")");
        if (completed) return;
        round = 5;
        lastAnnounced = 5;
        doneAtMs = System.currentTimeMillis();
        announceRound(5);
        completed = true;
    }

    /** Center-screen title hook (titles are local-only, so a loose match is safe). */
    public static void onTitle(String title) {
        if (!FishSettings.simonSaysEnabled || title == null) return;
        String s = title.replaceAll("§.", "").toLowerCase();
        if (s.contains("device") && s.contains("complete")) tryComplete();
    }

    private static void announceRound(int r) {
        String label = r + "/5" + (r >= 5 ? " (done)" : "");
        Misc.addChatMessage(Component.literal(fishmod.utils.FishMsg.prefix() + "§bSimon Says: §a" + label));
        if (FishSettings.simonSaysPartyChat) Misc.executeCommand("pc Simon Says: " + label);
    }

    // ── block scanning ──────────────────────────────────────────────────────────

    /**
     * Watches the fixed obsidian/button columns for a break, same signal NoammAddons uses.
     * Any obsidian cell missing = device active. Once that's held for {@code BREAK_COOLDOWN_TICKS}
     * and every button cell is air, the device reset — announce the fail, reset round tracking,
     * and go fully quiet (see {@code broken} in {@link #tick}) until the device is active again.
     */
    private static void tickBreakState(Level world) {
        breakTicks--;

        boolean active = false;
        BlockPos.MutableBlockPos m = new BlockPos.MutableBlockPos();
        outer:
        for (int y = DEV_Y_MIN; y <= DEV_Y_MAX; y++)
            for (int z = DEV_Z_MIN; z <= DEV_Z_MAX; z++) {
                m.set(DEV_OBSIDIAN_X, y, z);
                if (world.getBlockState(m).getBlock() != Blocks.OBSIDIAN) { active = true; break outer; }
            }

        if (active) {
            breakTicks = BREAK_COOLDOWN_TICKS;
            canBreak = true;
            if (broken) {
                broken = false;
                if (debug) log("device restarted — resuming");
            }
            return;
        }

        if (breakTicks > 0 || !canBreak) return;

        boolean allAir = true;
        outer2:
        for (int y = DEV_Y_MIN; y <= DEV_Y_MAX; y++)
            for (int z = DEV_Z_MIN; z <= DEV_Z_MAX; z++) {
                m.set(DEV_BUTTONS_X, y, z);
                if (world.getBlockState(m).getBlock() != Blocks.AIR) { allAir = false; break outer2; }
            }
        if (!allAir) return;

        canBreak = false;
        broken = true;
        round = 0; maxLen = 0; lastAnnounced = 0; burstFlashes = 0;
        if (debug) log("device broke — reset + fully off until restart");

        if (FishSettings.simonSaysFailEnabled) {
            Misc.addChatMessage(Component.literal(fishmod.utils.FishMsg.prefix() + "§c" + FishSettings.simonSaysFailMessage));
            if (FishSettings.simonSaysPartyChat) Misc.executeCommand("pc " + FishSettings.simonSaysFailMessage);
        }
    }

    /** Fills {@code litOut} with lit sea-lantern positions in the box around the locked device center. */
    private static void scanLitCells(Level world, BlockPos center, java.util.HashSet<Long> litOut) {
        int cx = center.getX(), cy = center.getY(), cz = center.getZ();
        BlockPos.MutableBlockPos m = new BlockPos.MutableBlockPos();
        for (int dx = -SCAN_RADIUS; dx <= SCAN_RADIUS; dx++)
            for (int dy = -SCAN_RADIUS; dy <= SCAN_RADIUS; dy++)
                for (int dz = -SCAN_RADIUS; dz <= SCAN_RADIUS; dz++) {
                    m.set(cx + dx, cy + dy, cz + dz);
                    if (world.getBlockState(m).getBlock() == Blocks.SEA_LANTERN) litOut.add(m.asLong());
                }
    }

    /** Phase.inP3() but never throws — blade-addons' Phase may differ; treat errors as false. */
    private static boolean safeInP3() {
        try { return Phase.inP3(); } catch (Throwable t) { return false; }
    }

    private static void reset() {
        round = 0;
        maxLen = 0;
        lastAnnounced = 0;
        burstFlashes = 0;
        primed = false;
        completed = false;
        armed = false;
        breakTicks = 0;
        canBreak = false;
        broken = false;
        doneAtMs = 0L;
        litPrev.clear();
        deviceCenter = null;
    }

    public static int getStage() { return round; }

    public static void renderHud(GuiGraphicsExtractor ctx, DeltaTracker tc) {
        if (!FishSettings.simonSaysEnabled || !FishSettings.simonSaysHudEnabled) return;
        if (round <= 0) return;
        // Auto-hide 2 seconds after completion.
        if (round >= 5 && doneAtMs > 0 && System.currentTimeMillis() - doneAtMs > 2000) return;
        Minecraft mc = Minecraft.getInstance();
        if (mc.player == null) return;
        // Don't render when standing right at the device (~3 blocks) — you can see it yourself.
        if (deviceCenter != null && mc.player.blockPosition().distSqr(deviceCenter) <= 12) return;

        String label = round == 0
                ? "§bSimon Says: §7—"
                : "§bSimon Says: §a" + round + "§7/5" + (round >= 5 ? " §7(done)" : "");

        float sc = (float) FishSettings.simonSaysHudScale;
        ctx.pose().pushMatrix();
        ctx.pose().translate((float) FishSettings.simonSaysHudX, (float) FishSettings.simonSaysHudY);
        ctx.pose().scale(sc, sc);
        ctx.text(mc.font, label, 0, 0, -1, true);
        ctx.pose().popMatrix();
    }

    private static void log(String line) {
        fishmod.utils.debug.Debug.LOGGER.info("[SS] {}", line);
        Misc.addChatMessage(Component.literal("§e[SS] " + line));
    }

    /** /ssdbg: logs every block transition in a cube around the player — for locating the device. */
    private static void debugTick(Minecraft client) {
        if (!debug || client.player == null || client.level == null) { dbgPrev.clear(); return; }
        Level world = client.level;
        BlockPos c = client.player.blockPosition();
        BlockPos.MutableBlockPos m = new BlockPos.MutableBlockPos();
        boolean first = dbgPrev.isEmpty();
        for (int dx = -DBG_R; dx <= DBG_R; dx++)
            for (int dy = -DBG_R; dy <= DBG_R; dy++)
                for (int dz = -DBG_R; dz <= DBG_R; dz++) {
                    m.set(c.getX() + dx, c.getY() + dy, c.getZ() + dz);
                    net.minecraft.world.level.block.Block b = world.getBlockState(m).getBlock();
                    long key = m.asLong();
                    net.minecraft.world.level.block.Block old = dbgPrev.put(key, b);
                    if (first || old == b) continue;
                    String oldId = old == null ? "none" : net.minecraft.core.registries.BuiltInRegistries.BLOCK.getKey(old).getPath();
                    String newId = net.minecraft.core.registries.BuiltInRegistries.BLOCK.getKey(b).getPath();
                    log("(" + dx + "," + dy + "," + dz + ") " + oldId + " -> " + newId);
                }
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/dungeon/SimonSaysTracker.kt
package fishmod.features.dungeon

import fishmod.features.FishHudEditor
import fishmod.utils.Misc
import fishmod.utils.config.values.FishSettings
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.fabricmc.fabric.api.client.rendering.v1.hud.HudElementRegistry
import net.minecraft.client.DeltaTracker
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.core.BlockPos
import net.minecraft.resources.Identifier
import net.minecraft.network.chat.Component
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.AABB
import java.util.regex.Pattern

object SimonSaysTracker {


    private val DEVICE_BOX = AABB(
        106.65, 120.0, 92.70,
        110.70, 126.0, 95.30
    )
    private val DEVICE_CENTER = BlockPos(108, 120, 94)

    private const val DEV_BUTTONS_X = 110
    private const val DEV_OBSIDIAN_X = 111
    private const val DEV_Y_MIN = 120
    private const val DEV_Y_MAX = 123
    private const val DEV_Z_MIN = 92
    private const val DEV_Z_MAX = 95
    private const val BREAK_COOLDOWN_TICKS = 12

    private const val BREAK_GRACE_MS = 6000L

    private var round = 0
    private var resetAtMs = 0L
    private var lastAnnounced = 0
    private var primed = false
    private var completed = false
    private var completeLatched = false
    private var armed = false
    private var breakTicks = 0
    private var canBreak = false
    private var broken = false
    private var falseFailSent = false
    private var breakArmedAtMs = 0L
    private var atDevice = false
    private var deviceCenter: BlockPos? = null
    private var doneAtMs = 0L
    private val litPrev = HashSet<Long>()
    private val scanBuf = HashSet<Long>()
    private var seqLen = 0
    private var skipOver = false
    private var buttonsUp: Boolean? = null
    private val BUTTON_CHECK = BlockPos(110, 120, 93)

    private val SS_CHAT: Pattern = Pattern.compile("Simon Says: (\\d)/5")

    private const val GOLDOR_INTRO = "who dares trespass into my domain"

    @JvmField
    var debug = false
    private const val DBG_R = 6
    private val dbgPrev = HashMap<Long, Block?>()

    @JvmStatic
    fun init() {
        fishmod.utils.events.Events.ON_LOCATION_CHANGE.register { reset(); false }

        fishmod.utils.events.Events.ON_GAME_MESSAGE.register { message ->
            if (!FishSettings.simonSaysEnabled) return@register false
            val s = message.string.replace(fishmod.utils.Constants.STRIP_COLOR_REGEX, "")

            if (!armed && s.lowercase().contains(GOLDOR_INTRO)) {
                armed = true
                if (debug) log("armed (Goldor intro seen)")
            }

            if (!completeLatched) {
                val ss = SS_CHAT.matcher(s)
                if (ss.find()) {
                    val n = ss.group(1)[0] - '0'
                    if (n in 1..4) {
                        round = n
                    } else if (n >= 5) {
                        round = 5
                        doneAtMs = System.currentTimeMillis()
                        if (!completed && falseFailSent) {
                            Misc.addChatMessage(Component.literal(fishmod.utils.FishMsg.prefix() + "§a(actually completed — ignore the FAILED above)"))
                        }
                        completed = true
                        completeLatched = true
                        breakArmedAtMs = 0L
                        canBreak = false
                        broken = false
                        falseFailSent = false
                    }
                }
            }

            if (debug && s.contains("device")) log("msg: \"$s\"")
            if (!s.contains("completed a device")) return@register false
            val mc = Minecraft.getInstance()
            val self = mc.player?.gameProfile?.name()
            val mine = (self == null) || s.contains(self)
            if (debug) log("completed-a-device match; self=$self mine=$mine completed=$completed")
            if (mine) tryComplete()
            false
        }

        ClientTickEvents.END_CLIENT_TICK.register(ClientTickEvents.EndTick { debugTick(it); tick(it) })
        HudElementRegistry.addLast(Identifier.fromNamespaceAndPath("fishmod", "simon_says_tracker")) { ctx, tc -> if (!fishmod.features.FishHudEditor.isOpen()) renderHud(ctx, tc) }

        FishHudEditor.register(
            "Simon Says",
            { FishSettings.simonSaysHudX }, { v -> FishSettings.simonSaysHudX = v },
            { FishSettings.simonSaysHudY }, { v -> FishSettings.simonSaysHudY = v },
            110, 14,
            { FishSettings.simonSaysHudScale }, { v -> FishSettings.simonSaysHudScale = v },
            { FishSettings.simonSaysHudEnabled }
        )
    }

    private fun tick(client: Minecraft) {
        if (!FishSettings.simonSaysEnabled || client.player == null || client.level == null) {
            atDevice = false; deviceCenter = null; reset(); return
        }

        if (completed) { atDevice = false; return }

        if (!armed) { atDevice = false; return }

        tickBreakState(client.level!!)
        if (broken) {
            atDevice = false; deviceCenter = null; primed = false; litPrev.clear(); seqLen = 0; buttonsUp = null
            return
        }

        atDevice = false
        for (p: Player in client.level!!.players()) {
            if (p.boundingBox.intersects(DEVICE_BOX)) { atDevice = true; break }
        }

        if (!atDevice) {
            deviceCenter = null; primed = false; litPrev.clear()
            return
        }

        if (deviceCenter == null) {
            deviceCenter = DEVICE_CENTER
            primed = false; seqLen = 0; skipOver = false; litPrev.clear(); buttonsUp = null
            if (debug) log("locked device center " + deviceCenter!!.toShortString())
        }

        val world = client.level!!
        if (!world.hasChunk(DEV_OBSIDIAN_X shr 4, DEV_Z_MIN shr 4)) return

        // NoammAddons approach: count lanterns lighting on the 4x4 grid, lock the stage in when the buttons come back.
        scanBuf.clear()
        val m = BlockPos.MutableBlockPos()
        for (y in DEV_Y_MIN..DEV_Y_MAX) for (z in DEV_Z_MIN..DEV_Z_MAX) {
            m.set(DEV_OBSIDIAN_X, y, z)
            if (world.getBlockState(m).block == Blocks.SEA_LANTERN) scanBuf.add(m.asLong())
        }
        if (!primed) { litPrev.clear(); litPrev.addAll(scanBuf); primed = true }
        else {
            for (p in scanBuf) if (!litPrev.contains(p)) {
                seqLen = (seqLen + 1).coerceAtMost(5)
                if (FishSettings.simonSaysSkipCompat && seqLen == 2 && !skipOver) seqLen--
                if (debug) log("flash seqLen=$seqLen")
            }
            litPrev.clear(); litPrev.addAll(scanBuf)
        }

        val btn = world.getBlockState(BUTTON_CHECK).block
        val up = btn == Blocks.STONE_BUTTON
        if (btn == Blocks.AIR) seqLen = 0
        if (up && buttonsUp == false) {
            skipOver = true
            if (seqLen > 0 && seqLen > round && seqLen < 5) {
                round = seqLen
                if (round > lastAnnounced) { lastAnnounced = round; announceRound(round) }
            }
        }
        buttonsUp = up
    }

    private fun tryComplete() {
        if (debug) log("tryComplete called (completed=$completed)")
        if (completed) return
        round = 5
        lastAnnounced = 5
        doneAtMs = System.currentTimeMillis()
        announceRound(5)
        if (falseFailSent) {
            Misc.addChatMessage(Component.literal(fishmod.utils.FishMsg.prefix() + "§a(actually completed — ignore the FAILED above)"))
            if (FishSettings.simonSaysPartyChat) fishmod.utils.ChatQueue.enqueue("pc Simon Says: actually completed, ignore the FAILED above")
        }
        completed = true
        completeLatched = true
        breakArmedAtMs = 0L
        canBreak = false
        broken = false
        falseFailSent = false
    }

    @JvmStatic
    fun onTitle(title: String?) {
        if (!FishSettings.simonSaysEnabled || title == null) return
        val s = title.replace(fishmod.utils.Constants.STRIP_COLOR_REGEX, "").lowercase()
        if (s.contains("device") && s.contains("complete")) tryComplete()
    }

    private fun announceRound(r: Int) {
        val label = "$r/5" + (if (r >= 5) " (done)" else "")
        Misc.addChatMessage(Component.literal(fishmod.utils.FishMsg.prefix() + "§bSimon Says: §a" + label))
        val (custom, text) = when (r) {
            1 -> FishSettings.simon1Enabled to FishSettings.simon1Message
            2 -> FishSettings.simon2Enabled to FishSettings.simon2Message
            3 -> FishSettings.simon3Enabled to FishSettings.simon3Message
            4 -> FishSettings.simon4Enabled to FishSettings.simon4Message
            else -> FishSettings.simon5Enabled to FishSettings.simon5Message
        }
        when {
            custom && text.isNotBlank() -> fishmod.utils.ChatQueue.enqueue("pc $text")
            FishSettings.simonSaysPartyChat -> fishmod.utils.ChatQueue.enqueue("pc Simon Says: $label")
        }
    }

    private fun tickBreakState(world: Level) {
        if (completeLatched) { breakArmedAtMs = 0L; return }
        if (!world.hasChunk(DEV_OBSIDIAN_X shr 4, DEV_Z_MIN shr 4)) { breakArmedAtMs = 0L; return }

        breakTicks--

        var active = false
        val m = BlockPos.MutableBlockPos()
        outer@ for (y in DEV_Y_MIN..DEV_Y_MAX)
            for (z in DEV_Z_MIN..DEV_Z_MAX) {
                m.set(DEV_OBSIDIAN_X, y, z)
                if (world.getBlockState(m).block != Blocks.OBSIDIAN) { active = true; break@outer }
            }

        if (active) {
            breakTicks = BREAK_COOLDOWN_TICKS
            canBreak = true
            breakArmedAtMs = 0L
            if (broken) {
                broken = false
                if (debug) log("device restarted — resuming")
            }
            return
        }

        if (breakTicks > 0 || !canBreak) return

        var allAir = true
        outer2@ for (y in DEV_Y_MIN..DEV_Y_MAX)
            for (z in DEV_Z_MIN..DEV_Z_MAX) {
                m.set(DEV_BUTTONS_X, y, z)
                if (world.getBlockState(m).block != Blocks.AIR) { allAir = false; break@outer2 }
            }
        if (!allAir) { breakArmedAtMs = 0L; return }

        val now = System.currentTimeMillis()
        if (breakArmedAtMs == 0L) { breakArmedAtMs = now; return }

        val sinceClick = now - fishmod.features.dungeon.f7.SimonSaysSolver.lastRoundCompleteMs
        if (fishmod.features.dungeon.f7.SimonSaysSolver.lastRoundCompleteMs != 0L && sinceClick in 0..BREAK_GRACE_MS) {
            tryComplete()
            return
        }

        if (now - breakArmedAtMs < BREAK_GRACE_MS) return

        canBreak = false
        broken = true
        breakArmedAtMs = 0L
        round = 0; lastAnnounced = 0
        resetAtMs = now
        if (debug) log("device broke — reset + fully off until restart")

        if (FishSettings.simonSaysFailEnabled) {
            falseFailSent = true
            Misc.addChatMessage(Component.literal(fishmod.utils.FishMsg.prefix() + "§c" + FishSettings.simonSaysFailMessage))
            if (FishSettings.simonSaysPartyChat) fishmod.utils.ChatQueue.enqueue("pc " + FishSettings.simonSaysFailMessage)
        }
    }

    private fun reset() {
        round = 0
        lastAnnounced = 0
        primed = false
        completed = false
        completeLatched = false
        armed = false
        breakTicks = 0
        canBreak = false
        broken = false
        breakArmedAtMs = 0L
        doneAtMs = 0L
        falseFailSent = false
        litPrev.clear()
        deviceCenter = null
        seqLen = 0; skipOver = false; buttonsUp = null
    }

    @JvmStatic
    fun renderHud(ctx: GuiGraphicsExtractor, tc: DeltaTracker) {
        if (!FishSettings.simonSaysEnabled || !FishSettings.simonSaysHudEnabled) return
        val now = System.currentTimeMillis()
        val showReset = round <= 0 && FishSettings.ssProgressShowReset && resetAtMs > 0 && now - resetAtMs < 3000
        if (round <= 0 && !showReset) return
        if (round >= 5 && doneAtMs > 0 && now - doneAtMs > 2000) return
        val mc = Minecraft.getInstance()
        val player = mc.player ?: return
        val dc = deviceCenter
        if (dc != null && player.blockPosition().distSqr(dc) <= 12) return

        val (text, color) = when {
            showReset -> FishSettings.ssProgressResetText to FishSettings.ssProgressResetColor
            round >= 5 -> {
                if (!FishSettings.ssProgressShowCompleted) return
                FishSettings.ssProgressCompletedText to FishSettings.ssProgressCompletedColor
            }
            else -> {
                if (!FishSettings.ssProgressShowProgress) return
                FishSettings.ssProgressProgressText.replace("(n)", round.toString()) to FishSettings.ssProgressProgressColor
            }
        }
        val label = Component.literal(text).withColor(color and 0xFFFFFF)

        val sc = FishSettings.simonSaysHudScale.toFloat()
        ctx.pose().pushMatrix()
        ctx.pose().translate(FishSettings.simonSaysHudX.toFloat(), FishSettings.simonSaysHudY.toFloat())
        ctx.pose().scale(sc, sc)
        ctx.text(mc.font, label, 0, 0, -1, true)
        ctx.pose().popMatrix()
    }

    private fun log(line: String) {
        fishmod.utils.debug.Debug.LOGGER.info("[SS] {}", line)
        Misc.addChatMessage(Component.literal("§e[SS] $line"))
    }

    private fun debugTick(client: Minecraft) {
        if (!debug || client.player == null || client.level == null) { dbgPrev.clear(); return }
        val world = client.level!!
        val c = client.player!!.blockPosition()
        val m = BlockPos.MutableBlockPos()
        val first = dbgPrev.isEmpty()
        for (dx in -DBG_R..DBG_R)
            for (dy in -DBG_R..DBG_R)
                for (dz in -DBG_R..DBG_R) {
                    m.set(c.x + dx, c.y + dy, c.z + dz)
                    val b = world.getBlockState(m).block
                    val key = m.asLong()
                    val old = dbgPrev.put(key, b)
                    if (first || old === b) continue
                    val oldId = if (old == null) "none" else net.minecraft.core.registries.BuiltInRegistries.BLOCK.getKey(old).path
                    val newId = net.minecraft.core.registries.BuiltInRegistries.BLOCK.getKey(b).path
                    log("($dx,$dy,$dz) $oldId -> $newId")
                }
    }
}
